### PR TITLE
Bump to apimachinery 1.26

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

 * Bump all apimachinery references to 1.26
 * Requires an openapi-gen bump to make use of the new k8s.io/kube-openapi from 1.26
 * Adjusts informer registration callback registration calls with new error return signature
 * Adjust kubernetes Update call signature
 * Leader-election: Switch from `endpoints` to `leases`, since `endpoints` are deprecated
 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
Bump to apimachinery 1.26
```

### CODE REVIEW

File: pkg/virt-operator/resource/apply/admissionregistration_test.go [hunk 1]
```diff
@@ -127,19 +127,8 @@ var _ = Describe("WebhookConfiguration types", func() {
...
-			webhookv1beta1 := &admissionregistrationv1beta1.MutatingWebhookConfiguration{}
-			err = webhookv1beta1.Unmarshal(binv1)
-			Expect(err).ToNot(HaveOccurred())
-			Expect(webhookv1beta1.String()).To(Equal(webhook.String()))
-			binv1beta1, err := webhookv1beta1.Marshal()
-			Expect(err).ToNot(HaveOccurred())
-			Expect(binv1beta1).To(Equal(binv1))
 
-			webhookv1beta1, err = convertV1MutatingWebhookToV1beta1(webhook)
+			webhookv1beta1, err := convertV1MutatingWebhookToV1beta1(webhook)
```
Comment (xpivarc) [inline pkg/virt-operator/resource/apply/admissionregistration_test.go:131]: Do we actually need this anymore?

File: tests/operator/operator.go [hunk 2]
```diff
@@ -2677,7 +2676,7 @@ spec:
 					}},
 				},
 			}
-			errMsg := fmt.Sprintf("Unsupported value: \"%s\"", incorrectOperator)
+			errMsg := "spec.infra.nodePlacement.tolerations.operator in body should be one of"
```
Comment (xpivarc) [inline tests/operator/operator.go:2679]: How is this related?

**Review by xpivarc — COMMENTED** [review]
LGTM

File: staging/src/kubevirt.io/client-go/go.mod [hunk 3]
```diff
@@ -5,100 +5,98 @@ go 1.17
...
+	github.com/onsi/gomega v1.23.0
 	github.com/openshift/client-go v0.0.0-20210112165513-ebc401615f47
 	github.com/spf13/pflag v1.0.5
-	k8s.io/api v0.23.5
-	k8s.io/apiextensions-apiserver v0.23.5
-	k8s.io/apimachinery v0.23.5
+	k8s.io/api v0.26.3
+	k8s.io/apiextensions-apiserver v0.26.3
+	k8s.io/apimachinery v0.26.3
 	k8s.io/client-go v12.0.0+incompatible
```
Comment (lion7) [inline staging/src/kubevirt.io/client-go/go.mod:21]: Shouldn't this be bumped to v0.26.3 as well?

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/go.mod:21]: It is bumped in the replace section below. I can bump it here too (and will do, thanks), but from a functionality perspective it is ok.

**Comment by xpivarc** [comment]
> > > Leader-election: Switch from endpoints to leases, since endpoints are removed
> > 
> > 
> > /hold
> > Putting a hold here since we may require some logic to avoid a split-brain situation during updates.
> 
> @xpivarc any thoughts on this one? I thought about explicitly querying for the endpoints named `virt-operator` and `virt-controller` and only proceed to the new leader election mechanism if we see that the leader-election annotation, like here:
> 
> ```yaml
> apiVersion: v1
> kind: Endpoints
> metadata:
>   annotations:
>     control-plane.alpha.kubernetes.io/leader: '{"holderIdentity":"virt-controller-b4889694b-m27gb","leaseDurationSeconds":15,"acquireTime":"2022-11-17T09:36:52Z","renewTime":"2023-05-08T15:45:35Z","leaderTransitions":0}'
>   creationTimestamp: "2022-11-17T09:36:52Z"
>   name: virt-controller
>   namespace: vm-system
>   resourceVersion: "236089021"
>   uid: 0a6371e2-ce06-4147-9d4c-7048ed4dcb64
> ```
> 
> shows that we can proceed.

I am not quite sure what is the worry. You are using `EndpointsLeasesResourceLock` which is `MultiLock` and as far I see it should transparently handle the transition. What am I missing?

**Comment by rmohr** [comment]
> I am not quite sure what is the worry. You are using `EndpointsLeasesResourceLock` which is `MultiLock` and as far I see it should transparently handle the transition. What am I missing?

Oh you are right. Great!

/unhold

whil push a rebase.

**Comment by xpivarc** [comment]
@fossedihelm 
Can you have a look here once rebased, please?

**Comment by fossedihelm** [comment]
@rmohr Thanks! Looks good to me! Once you push the rebase I'll be happy to lgtm

**Comment by rmohr** [comment]
@fossedihelm @xpivarc finally managed to do the rebase. PTAL. Thanks!

File: pkg/storage/snapshot/restore_test.go [hunk 4]
```diff
@@ -1398,7 +1407,13 @@ func expectPVCCreateWithDataSourceRef(client *k8sfake.Clientset, vmRestore *snap
 
 		Expect(createObj.Spec.DataSource).ToNot(BeNil())
 		Expect(createObj.Spec.DataSourceRef).ToNot(BeNil())
-		Expect(createObj.Spec.DataSource).To(Equal(createObj.Spec.DataSourceRef))
+
+		dataSourceRef := &corev1.TypedLocalObjectReference{
```
Comment (fossedihelm) [inline pkg/storage/snapshot/restore_test.go:1411]: nit: `dataSource` instead of `dataSourceRef`

File: pkg/virt-api/api.go [hunk 5]
```diff
@@ -723,10 +727,15 @@ func (app *virtAPIApp) Compose() {
 }
 
 func (app *virtAPIApp) ConfigureOpenAPIService() {
-	spec := openapi.LoadOpenAPISpec(restful.RegisteredWebServices())
-	config := openapi.CreateOpenAPIConfig(restful.RegisteredWebServices())
+	config := openapi.CreateConfig()
+	config.GetDefinitions = v12.GetOpenAPIDefinitions
+	spec, err := builderv3.BuildOpenAPISpec(restful.RegisteredWebServices(), config)
```
Comment (fossedihelm) [inline pkg/virt-api/api.go:732]: nit: usage of deprecated function
```suggestion
        spec, err := builderv3.BuildOpenAPISpecFromRoutes(restfuladapter.AdaptWebServices(restful.RegisteredWebServices()), config)
```

**Review by fossedihelm — COMMENTED** [review]
Thanks @rmohr! Two small nits that can be addressed in future!
/lgtm

File: pkg/virt-api/api.go (hunk 5 above)
Comment (rmohr) [reply pkg/virt-api/api.go:732]: Yes good catch, I tried to get away from it, but it is not such an easy task and would require quite a bit more openapi work.

File: pkg/storage/snapshot/restore_test.go (hunk 4 above)
Comment (fossedihelm) [reply pkg/storage/snapshot/restore_test.go:1411]: Well, intricate :) You are right, since both `Spec.DataSource` and `Spec.DataSourceRef` are references. My reasoning was mainly based on the fact that you are then going to compare it with `Spec.DataSource`.
It can stay as is from my point of view

File: pkg/virt-api/api.go (hunk 5 above)
Comment (fossedihelm) [reply pkg/virt-api/api.go:732]: IMHO it can be addresses in a follow up PR, to allow this to go forward. Thanks
